core/chains/evm/logpoller/parser.go

```diff
@@ -0,0 +1,528 @@
+	sorting := limiter.SortBy
+
+	if limiter.HasCursorLimit() && !limiter.HasSequenceSort() {
+		var dir query.SortDirection
+
+		switch limiter.Limit.CursorDirection {
+		case query.CursorFollowing:
+			dir = query.Asc
+		case query.CursorPrevious:
+			dir = query.Desc
+		default:
+			return "", errors.New("unexpected cursor direction")
+		}
+
+		sorting = append(sorting, query.NewSortBySequence(dir))
+	}
+
+	if len(sorting) == 0 {
+		return "", nil
+	}
+
+	sort := make([]string, len(sorting))
+
+	for idx, sorted := range sorting {
+		var name string
+
+		order, err := orderToString(sorted.GetDirection())
+		if err != nil {
+			return "", err
+		}
+
+		switch sorted.(type) {
+		case query.SortByBlock:
+			name = blockFieldName
+		case query.SortBySequence:
+			sort[idx] = fmt.Sprintf("block_number %s, tx_hash %s, log_index %s", order, order, order)
+
+			continue
+		case query.SortByTimestamp:
+			name = timestampFieldName
+		default:
+			return "", errors.New("unexpected sort by")
+		}
+
+		sort[idx] = fmt.Sprintf("%s %s", name, order)
+	}
+
+	return fmt.Sprintf("ORDER BY %s", strings.Join(sort, ", ")), nil
+}
+
+func (v *pgDSLParser) limitClause(limiter query.LimitAndSort) string {
+	if !limiter.HasCursorLimit() && limiter.Limit.Count == 0 {
+		return ""
+	}
+
+	return fmt.Sprintf("LIMIT %d", limiter.Limit.Count)
+}
+
+func (v *pgDSLParser) getLastExpression() (string, error) {
+	exp := v.expression
+	err := v.err
+
+	v.expression = ""
+	v.err = nil
+
+	return exp, err
+}
+
+func (v *pgDSLParser) combineExpressions(expressions []query.Expression, op query.BoolOperator) (string, error) {
+	grouped := len(expressions) > 1
+	clauses := make([]string, len(expressions))
+
+	for idx, exp := range expressions {
+		if exp.IsPrimitive() {
+			exp.Primitive.Accept(v)
+
+			clause, err := v.getLastExpression()
+			if err != nil {
+				return "", err
+			}
+
+			clauses[idx] = clause
+		} else {
+			clause, err := v.combineExpressions(exp.BoolExpression.Expressions, exp.BoolExpression.BoolOperator)
+			if err != nil {
+				return "", err
+			}
+
+			clauses[idx] = clause
+		}
+	}
+
+	output := strings.Join(clauses, fmt.Sprintf(" %s ", op.String()))
+
+	if grouped {
+		output = fmt.Sprintf("(%s)", output)
+	}
+
+	return output, nil
+}
+
+func cmpOpToString(op primitives.ComparisonOperator) (string, error) {
+	switch op {
+	case primitives.Eq:
+		return "=", nil
+	case primitives.Neq:
+		return "!=", nil
+	case primitives.Gt:
+		return ">", nil
+	case primitives.Gte:
+		return ">=", nil
+	case primitives.Lt:
+		return "<", nil
+	case primitives.Lte:
+		return "<=", nil
+	default:
+		return "", errors.New("invalid comparison operator")
+	}
+}
+
+func orderToString(dir query.SortDirection) (string, error) {
+	switch dir {
+	case query.Asc:
+		return "ASC", nil
+	case query.Desc:
+		return "DESC", nil
+	default:
+		return "", errors.New("invalid sort direction")
+	}
+}
+
+func valuesFromCursor(cursor string) (int64, string, int, error) {
```

Comment:
Some godoc or tests showing the expected cursor format would be very helpful here